lib/core/src/irods_service_account.cpp

```diff
@@ -0,0 +1,132 @@
+#include "irods/irods_service_account.hpp"
+
+#include "irods/irods_exception.hpp"
+#include "irods/irods_get_full_path_for_config_file.hpp"
+#include "irods/irods_kvp_string_parser.hpp"
+#include "irods/irods_log.hpp"
+#include "irods/rodsErrorTable.h"
+
+#include <cerrno>
+#include <cstdint>
+#include <cstdlib>
+#include <fstream>
+#include <memory>
+
+#include <pwd.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string.hpp>
+#include <fmt/format.h>
+
+namespace irods
+{
+    const std::string SERVICE_ACCOUNT_CONFIG_FILE("service_account.config");
+    const std::string SERTICE_ACCOUNT_KW("IRODS_SERVICE_ACCOUNT_NAME");
+
+    const size_t PWD_BUF_LEN_DEFAULT = 16384; // default size of buffer for strings in struct passwd
+    const int PWD_BUF_LEN_MINIMUM = 128; // minimum size of buffer for strings in struct passwd
+
+    // FIXME: Parsing of service_account.config isn't super robust.
+    auto is_service_account() -> bool
+    {
+        std::string service_account_username;
+
+        std::string svc_acct_cfg_fn;
+        irods::error ret = irods::get_full_path_for_config_file(SERVICE_ACCOUNT_CONFIG_FILE, svc_acct_cfg_fn);
+
+        if (!ret.ok()) {
+            // if config file not found, the machine probably does not have a server
+            return false;
+        }
+
+        std::ifstream svc_acct_cfg_strm{svc_acct_cfg_fn};
+        if (!svc_acct_cfg_strm.is_open()) {
+            // if we can't open the config file, it's likely due to a permission issue
+            // meaning it is extremely unlikely we are the service account
+            return false;
+        }
+
+        kvp_map_t service_acct_cfg;
+        uint_fast64_t cfg_line_no = 0;
+        for (std::string cfg_line; std::getline(svc_acct_cfg_strm, cfg_line);) {
+            cfg_line_no++;
+            boost::algorithm::trim(cfg_line); // trim whitespace (should handle CRLF)
+            if (cfg_line.empty()) {
+                // ignore empty lines
+                continue;
+            }
+            if (cfg_line.at(0) == '#') {
+                // ignore lines starting with #
+                continue;
+            }
+
+            // For now, let's just re-use our kvp-parsing functionality
+            // FIXME: does not handle spaces around "=", quotes, inline comments, or escape-newline
+            ret = parse_escaped_kvp_string(cfg_line, service_acct_cfg);
+            if (!ret.ok()) {
+                irods::log(LOG_WARNING,
+                           fmt::format("{} - {} parse failure on line {}",
+                                       __PRETTY_FUNCTION__,
+                                       SERVICE_ACCOUNT_CONFIG_FILE,
+                                       cfg_line_no));
```

Comment:
Why not use the new logger?

Reply:
Does the new logger actually print to stdout/stderr? I thought it was only for the server

Reply:
It doesn't. Is that important for this function?

Reply:
I think it's reasonable for `iexit` to be able to let users know when we have trouble parsing `service_account.config`, or when the user specified in `service_account.config` doesn't seem to exist on their system. These aren't fatal errors, so the function needs some way to be able to indicate these problems to the user before moving on.

Reply:
Gotcha. It wasn't clear to me if this work was for the client or server or both.

Is this a function you imagine being usable from the server-side as well?

Reply:
Something to consider as well is that `irods::log()` and `rodsLog()` will be going away in main and 4-3-stable eventually.

Not a blocker at the moment though.

Reply:
> Is this a function you imagine being usable from the server-side as well?

Sure. It's a client-oriented function, but I wrote it in such a way that it could be called from the server if needed.

> Something to consider as well is that `irods::log()` and `rodsLog()` will be going away in main and 4-3-stable eventually.

I figured as much. Before that happens, we should have functionality for client-friendly logging though, right? Probably goes without saying, but I think it would be a bad idea to rip these out without something to replace them with.

Reply:
I don't think that is something we should provide. Of course, willing to discuss more and be persuaded otherwise as that time approaches.

> Probably goes without saying, but I think it would be a bad idea to rip these out without something to replace them with.

Absolutely. We're going to provide as much cushion for the users (that makes sense) with whatever is decided.

Reply:
Definitely a topic for discussion. I'll mark this as resolved.